sql/src/main/java/io/crate/execution/engine/window/WindowBatchIterator.java

```diff
@@ -0,0 +1,169 @@
+    private boolean peersCollected = false;
+    private Row prevRow = null;
+
+    WindowBatchIterator(WindowDefinition windowDefinition,
+                        BatchIterator<Row> source,
+                        Collector<Row, ?, Iterable<Row>> collector) {
+        assert windowDefinition.partitions().size() == 0 : "Only empty OVER() is supported now";
+        assert windowDefinition.orderBy() == null : "Only empty OVER() is supported now";
+        assert windowDefinition.windowFrameDefinition() == null : "Only empty OVER() is supported now";
+
+        this.windowDefinition = windowDefinition;
+        this.source = source;
+        this.collector = collector;
+    }
+
+    @SuppressWarnings("unused")
+    private boolean arePeers(@Nullable Row prevRow, Row nextRow) {
+        // all rows are peers when orderBy is missing
+        return windowDefinition.orderBy() == null;
+    }
+
+    @Override
+    protected BatchIterator<Row> delegate() {
+        return source;
+    }
+
+    @Override
+    public Row currentElement() {
+        return currentElement;
+    }
+
+    @Override
+    public void moveToStart() {
+        super.moveToStart();
+        sourceRowsConsumed = 0;
+        windowRowPosition = 0;
+        peerRows = new ArrayList<>();
+        prevRow = null;
+        currentElement = null;
+    }
+
+    @Override
+    public boolean moveNext() {
+        if (peersCollected && windowRowPosition < sourceRowsConsumed) {
+            // emit the result of the window function as we computed the result and not yet emitted it for every row
+            // in the window
+            windowRowPosition++;
+            return true;
+        }
+
+        while (source.moveNext()) {
+            sourceRowsConsumed++;
+            Row sourceRow = source.currentElement();
+            if (arePeers(prevRow, sourceRow)) {
+                peersCollected = false;
+                RowN materializedRow = new RowN(sourceRow.materialize());
+                peerRows.add(materializedRow);
+                prevRow = materializedRow;
+            } else {
+                // rows are not peers anymore so compute the window function and emit the result
+                collectPeers();
+                windowRowPosition++;
+                return true;
+            }
+        }
+
+        if (source.allLoaded()) {
+            if (!peersCollected) {
+                collectPeers();
+            }
+
+            if (windowRowPosition < sourceRowsConsumed) {
+                // we still need to emit rows
+                windowRowPosition++;
+                return true;
+            } else {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private void collectPeers() {
+        BatchIterator<Row> peerRowsIterator = InMemoryBatchIterator.of(peerRows, SentinelRow.SENTINEL);
+        CollectingRowConsumer<?, Iterable<Row>> collectingRowConsumer = new CollectingRowConsumer<>(collector);
```

Comment:
We could avoid the BatchIterator/CollectingRowConsumer machinery here but directly operate on the `peerRows`, either via going through the stream API or by manually using the collector. 
(This would avoid the `resultFuture().join()` thing and make it clearer that we're operating on in-memory data here.)

Reply:
great idea